File: loan-management-microservices/loan-management-frontend/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '@environments/environment';
import { User, CreateUserRequest, UpdateUserRequest, UpdateUserRolesRequest } from '../models/user.model';
import { ApiResponse, PageResponse } from '../models/api-response.model';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private readonly API_URL = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) {}

  getAllUsers(page: number = 0, size: number = 10, sortBy: string = 'createdAt', sortDir: string = 'desc'): Observable<ApiResponse<PageResponse<User>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);

    return this.http.get<ApiResponse<PageResponse<User>>>(this.API_URL, { params });
  }

  getUserById(userId: number): Observable<ApiResponse<User>> {
    return this.http.get<ApiResponse<User>>(`${this.API_URL}/${userId}`);
  }

  getCurrentUser(): Observable<ApiResponse<User>> {
    return this.http.get<ApiResponse<User>>(`${this.API_URL}/me`);
  }

  /**
   * Update current logged-in user's profile
   */
  updateCurrentUser(request: UpdateUserRequest): Observable<ApiResponse<User>> {
    return this.http.put<ApiResponse<User>>(`${this.API_URL}/me`, request);
  }

  createUser(request: CreateUserRequest): Observable<ApiResponse<User>> {
    return this.http.post<ApiResponse<User>>(this.API_URL, request);
  }

  updateUser(userId: number, request: UpdateUserRequest): Observable<ApiResponse<User>> {
    return this.http.put<ApiResponse<User>>(`${this.API_URL}/${userId}`, request);
  }

  deactivateUser(userId: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.API_URL}/${userId}`);
  }

  activateUser(userId: number): Observable<ApiResponse<void>> {
    return this.http.patch<ApiResponse<void>>(`${this.API_URL}/${userId}/activate`, {});
  }

  updateUserRoles(userId: number, request: UpdateUserRolesRequest): Observable<ApiResponse<User>> {
    return this.http.put<ApiResponse<User>>(`${this.API_URL}/${userId}/roles`, request);
  }

  getAllRoles(): Observable<ApiResponse<string[]>> {
    return this.http.get<ApiResponse<string[]>>(`${this.API_URL}/roles`);
  }
}

File: loan-management-microservices/auth-service/src/main/java/com/loanmanagement/auth/domain/service/UserService.java
package com.loanmanagement.auth.domain.service;

import com.loanmanagement.auth.application.dto.request.CreateUserRequest;
import com.loanmanagement.auth.application.dto.request.UpdateUserRequest;
import com.loanmanagement.auth.application.dto.request.UpdateUserRolesRequest;
import com.loanmanagement.auth.application.dto.response.PageResponse;
import com.loanmanagement.auth.application.dto.response.UserResponse;
import com.loanmanagement.auth.domain.model.RoleType;
import com.loanmanagement.common.dto.UserDetailsDTO;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for user management operations
 */
public interface UserService {

    /**
     * Get all users (paginated)
     */
    PageResponse<UserResponse> getAllUsers(Pageable pageable);

    /**
     * Get user by ID
     */
    UserResponse getUserById(Long userId);

    /**
     * Get current logged-in user
     */
    UserResponse getCurrentUser();

    /**
     * Create new user (Admin/Loan Officer)
     */
    UserResponse createUser(CreateUserRequest request);

    /**
     * Update user
     */
    UserResponse updateUser(Long userId, UpdateUserRequest request);

    /**
     * Update user roles (Admin only)
     */
    UserResponse updateUserRoles(Long userId, UpdateUserRolesRequest request);

    /**
     * Deactivate user
     */
    void deactivateUser(Long userId);

    /**
     * Activate user
     */
    void activateUser(Long userId);

    /**
     * Get all available roles
     */
    List<RoleType> getAllRoles();

    /**
     * Check if user ID matches current user
     */
    boolean isCurrentUser(Long userId);

    /**
     * Get user details for inter-service communication
     */
    UserDetailsDTO getUserDetailsById(Long userId);

    /**
     * Get user details by username for inter-service communication
     */
    UserDetailsDTO getUserDetailsByUsername(String username);

    /**
     * Get multiple users details for inter-service communication
     */
    List<UserDetailsDTO> getUserDetailsByIds(List<Long> userIds);

    /**
     * Get total users count for reporting
     */
    Long getTotalUsersCount();

    /**
     * Get active users count for reporting
     */
    Long getActiveUsersCount();

    /**
     * Get users count by role for reporting
     */
    Long getUsersByRoleCount(String role);

    /**
     * Get user statistics for reporting
     */
    java.util.Map<String, Object> getUserStatistics();

    /**
     * Get all users for reporting (no pagination)
     */
    List<UserDetailsDTO> getAllUsers();
}
